pkg/controllermanager/controller/cloudprofile/cloudprofile_control.go

```diff
@@ -148,9 +154,30 @@ func (c *defaultControl) ReconcileCloudProfile(obj *gardencorev1beta1.CloudProfi
 	}
 
 	if err := controllerutils.EnsureFinalizer(context.TODO(), c.k8sGardenClient.Client(), cloudProfile, gardencorev1beta1.GardenerName); err != nil {
-		logger.Logger.Errorf("could not add finalizer to CloudProfile: %s", err.Error())
+		cloudProfileLogger.Errorf("could not add finalizer to CloudProfile: %s", err.Error())
 		return err
 	}
 
+	kubernetesVersionUpdateRequired := false
+	if c.config.KubernetesVersionManagement.Enabled {
+		if kubernetesVersionUpdateRequired, obj, err = kubernetesversion.ReconcileKubernetesVersions(cloudProfileLogger, c.config, cloudProfile); err != nil {
+			return err
+		}
+	}
+
+	machineImagesUpdateRequired := false
+	if c.config.MachineImageVersionManagement.Enabled {
+		if machineImagesUpdateRequired, obj, err = machineversion.ReconcileMachineImageVersions(cloudProfileLogger, c.config, cloudProfile); err != nil {
+			return err
+		}
+	}
+
+	if kubernetesVersionUpdateRequired || machineImagesUpdateRequired {
```

Comment:
Proposal: Deep copy cloud profile before calling the reconcile funcs and run `!equality.Semantic.DeepEqual(cloudProfile, cloudProfileCurrent)` here.